File: esercizio4/svoglimento/models/cluster.js
const Punto = require("./point");

module.exports = class Cluster {
    /**
     * Costruttore
     * @param {Array<Punto> | null}points
     * @constructor
     */
    constructor(...points) {
        this.points = [];
        if(points != null)
            this.points.concat(points);

        this._center = null;
    }

    /**
     * Aggiunte un punto al cluster
     * @param {Punto}point il punto da aggiungere
     */
    add(point) {
        this.points.push(point);
        this._center = null;
    }

    center() {
        if(this._center != null)
            return this._center;

        const size = this.points.length;
        if(size < 1)
            return new Punto(Infinity,Infinity);

        let sumX = 0;
        let sumY = 0;

        for(let i=0;i<size;i++) {
            sumX+=this.points[i].x;
            sumY+=this.points[i].y;
        }

        const center = new Punto(sumX/size, sumY/size);
        this._center = center;

        return center;
    }

    distance(other) {
        return other.center().distance(this.center());
    }

   /**
     * Raggruppa due cluster in uno nuovo
     * @param {module.Cluster}a
     * @param {module.Cluster}b
     * @returns {module.Cluster}
     */
    static union(a,b) {
        const newCluster = new Cluster();
        for(let i=0;i<a.points.length;i++) {
            newCluster.add(a.points[i]);
        }
        for(let i=0;i<b.points.length;i++) {
            newCluster.add(b.points[i]);
        }
        return newCluster;
    }

    toString() {
        return "{ x: "+this.center().x+", y: "+this.center().y+" }";
    }

    /**
     * Ritorna l'errore del clusters
     * @returns {number}
     */
    get error() {
        let totalError = 0;
        for(let i=0;i<this.points.length;i++) {
            const contea = this.points[i];
            const delta = contea.distance(this.center());
            totalError += (contea.popolazione * Math.pow(delta,2));
        }

        return totalError;
    }

    /**
     * Ritorna la distorsione di un array di cluster
     * @param {Array<Cluster>}clusters il soggetto
     * @returns {number}
     */
    static getDistortion(clusters) {
        let totalDistortion = 0;
        for(let i=0;i<clusters.length;i++) {
            totalDistortion += clusters[i].error;
        }
        return totalDistortion;
    }

    static cascadeDistortion(cascadeData) {
        const keys = Object.keys(cascadeData);
        const results = {};
        for(let i=0;i<keys.length;i++) {
            results[keys[i]] = Cluster.getDistortion(cascadeData[keys[i]]);
        }

        return results;
    }
};

File: esercizio5/src/it/homepc/tagliabuemichele/model/Cluster.java
package it.homepc.tagliabuemichele.model;

import java.util.ArrayList;
import java.util.List;

public class Cluster {
    private List<City> points;
    private Point center;

    public Cluster() {
        points = new ArrayList<>();
        center = null;
    }

    public Point center() {
        if(this.center != null)
            return this.center;

        int size = this.points.size();
        if(size < 1)
            return new Point(Double.POSITIVE_INFINITY,
                Double.POSITIVE_INFINITY);

        double sumLat = 0;
        double sumLng = 0;

        for(int i=0;i<size;i++) {
            sumLat+=points.get(i).getLat();
            sumLng+=points.get(i).getLng();
        }

        center = new Point(sumLat/size, sumLng/size);
        return center;
    }

    public double error() {
        double totalError = 0;

        for(int i=0;i<points.size();i++) {
            City city = points.get(i);
            double delta = city.distance(center());
            totalError += city.getPopulation() * Math.pow(delta,2);
        }

        return totalError;
    }

    public double distance(Cluster other) {
        return other.center().distance(this.center());
    }

    public void add(City point) {
        points.add(point);
    }

    @Override
    public String toString() {
        return "Cluster(centro: "+center()+", size: "+points.size()+")";
    }

    public static double distortion(List<Cluster> clusters) {
        double totalDistortion = 0;
        for(int i=0;i<clusters.size();i++) {
            totalDistortion += clusters.get(i).error();
        }
        return totalDistortion;
    }
}
